sdks/python/apache_beam/io/gcp/bigtable_io.py

```diff
@@ -0,0 +1,120 @@
+#
+# Licensed to the Apache Software Foundation (ASF) under one or more
+# contributor license agreements.  See the NOTICE file distributed with
+# this work for additional information regarding copyright ownership.
+# The ASF licenses this file to You under the Apache License, Version 2.0
+# (the "License"); you may not use this file except in compliance with
+# the License.  You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+"""BigTable connector
+
+This module implements writing to BigTable tables.
+The default mode is to set row data to write to BigTable tables.
+The syntax supported is described here:
+https://cloud.google.com/bigtable/docs/quickstart-cbt
+
+BigTable connector can be used as main outputs. A main output
+(common case) is expected to be massive and will be split into
+manageable chunks and processed in parallel. In the example below
+we created a list of rows then passed to the GeneratedDirectRows
+DoFn to set the Cells and then we call the WriteToBigtable to insert
+those generated rows in the table.
+
+  main_table = (p
+       | 'Generate Direct Rows' >> GenerateDirectRows(number)
+       | 'Write to BT' >> beam.ParDo(WriteToBigtable(config)))
+"""
+from __future__ import absolute_import
+
+import apache_beam as beam
+from apache_beam.metrics import Metrics
+from apache_beam.options.pipeline_options import PipelineOptions
+from apache_beam.transforms.display import DisplayDataItem
+
+try:
+  from google.cloud.bigtable import Client
+  from google.cloud.bigtable.batcher import MutationsBatcher
+except ImportError:
+  pass
+
+
+class WriteToBigtable(beam.DoFn):
+  """ Creates the connector can call and add_row to the batcher using each
+  row in beam pipe line
+
+  :type beam_options: class:`~bigtable_configuration.BigtableConfiguration`
+  :param beam_options: class `~bigtable_configuration.BigtableConfiguration`
+  """
+
+  def __init__(self, beam_options):
```

Comment:
Why can't we have a pickleable beam_options variable?

Reply:
Ah, I assumed beam_options to be a Beam pipeline options object that is expected to be used at pipeline construction to populate the transforms (not at runtime). Looks like it's not.

In any case, from an Python user API perspective, it might be better to get the parameters directly as keyword arguments instead of using a wrapper object. WDYT ?

Reply:
Ok, we need to set the information we set in the `BigtableConfiguration` in the pipeline options, and then access that data from the PTransform?

Reply:
My suggestion will be to make project_id, instance_id, and table_id keyword arguments to a "WriteToBigTable" ptransform and set these values as a picklable instance variables of a "_BigTableWriteFn" DoFn (with either a wrapper object or not). 

(There's no need to use a PipelineOptions object in this case.)

Reply:
Ok, I change set the project_id, instance_id and table_id in a dictionary to keep using the beam_options variable as a picklable instance variable.